### PR TITLE
add identity id in SF during contact merge

### PR DESCRIPTION
**Pull request by johnduffell**
During inital runs of the contact merge, everything worked fine apart from I forgot that identity id doesn't sync automatically zuora->sf.
This means that the SF->identity sync wasn't happening, also the identity link would be broken so breaking online management of products.
This is not OK so I have added that as an extra call.
This also removed the need to clear all the identity ids from zuora and add them after - as this will trigger no extra syncing.
The new order is to add the SF and identity link to all the zuora accounts.  Then clear the identity from the old SF (there is a unique constraint on identity id in SF).  Finally add the identity id where it should be in SF.  This will trigger the sync of billing info into identity from the winning SF account.

### CODE REVIEW

File: handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala [hunk 1]
```diff
@@ -64,35 +62,46 @@ object Handler {
...
+  //    lazy val value: A = a()
+  //  }
+  //  object Lazy {
+  //    def apply[A](a: => A): Lazy[A] = new Lazy(() => a)
+  //  }
+
   case class MergeRequest(
     sFPointer: SFPointer,
-    zuoraAccountIds: List[AccountId],
-    identityId: Option[IdentityId]
```
Comment (johnduffell) [inline handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala]: identity id is now in the first element of the mergerequest. maybe I need to go through the naming a bit.

File: handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala [hunk 2]
```diff
@@ -64,35 +62,46 @@ object Handler {
...
         SFContactId(request.fullContactId),
-        CRMAccountId(request.accountId)
+        CRMAccountId(request.accountId),
+        request.identityId.map(IdentityId.apply) // need to remove all identity ids and use this identity id on all the accounts after linking
       ),
-      request.billingAccountZuoraIds.map(AccountId.apply),
-      request.identityId.map(IdentityId.apply) // need to remove all identity ids and use this identity id on all the accounts after linking
+      accountIds
     )
+  }
```
Comment (johnduffell) [inline handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala]: this finction is now a type level validation of the request, however not one that checks against any external data sources

File: handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/update/UpdateAccountSFLinks.scala [hunk 3]
```diff
@@ -1,30 +1,35 @@
...
   )
 
   case class CRMAccountId(value: String) extends AnyVal
 
-  def apply(zuoraRequests: RequestsPUT)(sFPointer: SFPointer)(account: AccountId): ClientFailableOp[Unit] = {
-    val request = Request(sFPointer.crmAccountId.value, sFPointer.sfContactId.value)
-    val path = s"accounts/${account.value}" // TODO danger - we shoudn't go building urls with string concatenation!
-    zuoraRequests.put[Request, Unit](request, path)
+  def apply(put: PutRequest => ClientFailableOp[JsonResponse]): SFPointer => AccountId => ClientFailableOp[Unit] =
+    (toRequest _).andThen(_.andThen(put).andThen(_.map(_ => ())))
```
Comment (johnduffell) [inline handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/update/UpdateAccountSFLinks.scala]: this is a neat way of "prepending" and "postprocessing" a requst in a way that makes testing much nicer (no vars) however repeating this apply boilerplate everywhere won't be fun at all.  to do.

File: handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala [hunk 4]
```diff
@@ -64,35 +62,46 @@ object Handler {
...
+      mergeRequest <- wireRequestToDomainObject(wireInput).toApiGatewayContinueProcessing(ApiGatewayResponse.badRequest("no account ids supplied"))
+      accountAndEmails <- getIdentityAndZuoraEmailsForAccounts(mergeRequest.zuoraAccountIds)
+        .toApiGatewayOp("getIdentityAndZuoraEmailsForAccounts")
+      _ <- validation(mergeRequest.sFPointer.identityId, accountAndEmails)
+      oldContact = accountAndEmails.find(_.identityId.isDefined).map(_.sfContactId)
+      _ <- update(mergeRequest.sFPointer, oldContact, mergeRequest.zuoraAccountIds)
         .toApiGatewayOp("update accounts with winning details")
     } yield ApiGatewayResponse.successfulExecution).apiResponse
 
+  //  class Lazy[A](a: () => A) {
```
[the remark this answers is not in the document]
Comment (johnduffell) [reply handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala]: yes, good point, I was just experimenting ( @pvighi  had a similar issue on some other code so I left it to show him)

File: handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala [hunk 5]
```diff
@@ -64,35 +54,41 @@ object Handler {
...
-    identityId: Option[IdentityId]
+    sFPointer: LinksFromZuora,
+    zuoraAccountIds: NonEmptyList[AccountId]
   )
-  def wireRequestToDomainObject(request: Handler.WireSfContactRequest): MergeRequest =
-    MergeRequest(
-      SFPointer(
+
+  def apply(request: Handler.WireSfContactRequest): Option[MergeRequest] = {
+    for {
```
Comment (pvighi) [inline handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala]: does this have to be a for comprehension? or can it just me a map?

File: handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/update/UpdateSalesforceIdentityId.scala [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import com.gu.util.resthttp.Types.ClientFailableOp
+import play.api.libs.json.Json
+
+object UpdateSalesforceIdentityId {
+
+  case class WireRequest(IdentityID__c: String)
+  implicit val writes = Json.writes[WireRequest]
+
+  def apply(patch: PatchRequest => ClientFailableOp[Unit]): ((SFContactId, Option[IdentityId])) => ClientFailableOp[Unit] =
+    (toRequest _).tupled.andThen(patch)
```
Comment (pvighi) [inline handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/update/UpdateSalesforceIdentityId.scala]: 👍

**Comment by pvighi** [comment]
I'm not sure if you're done with it  but looks good 👍

**Pull request merged**